compiler/util/kernel_type.py

```diff
@@ -0,0 +1,240 @@
+
+        # check if op is a multiplication
+        if not isinstance(op, arith.Muli):
+            raise KernelException("not a multiplication")
+
+        # get operands
+        a = op.lhs
+        b = op.rhs
+        return a, b
+
+    @staticmethod
+    def parse_add(op: Operation) -> tuple[Operation | SSAValue, Operation | SSAValue]:
+        """
+        Parses an addition operation and returns the two operands.
+
+        Args:
+            op (Operation): The addition operation.
+
+        Returns:
+            tuple[Operation | SSAValue, Operation | SSAValue]: The two operands
+            of the addition operation.
+
+        Raises:
+            KernelException: If the provided operation is not an addition.
+        """
+
+        # check if op is a multiplication
+        if not isinstance(op, arith.Addi):
+            raise KernelException("not an addition")
+
+        # get operands
+        a = op.lhs
+        b = op.rhs
+        return a, b
+
+    @staticmethod
+    def parse_zpa(op: Operation) -> tuple[Operation | SSAValue, Operation | SSAValue]:
+        """
+        Parses a zero point adjustment operation and returns the two operands.
+        A zero point adjustment operation is a subtraction operation, where
+        the right operand is a zero point offset and the left operand is the
+        input tensor. Before the subtraction, the input tensor may be sign
+        extended to avoid overflow.
+
+        Args:
+            op (Operation): The subtraction operation.
+
+        Returns:
+            tuple[Operation | SSAValue, Operation | SSAValue]: The two operands
+            of the subtraction operation.
+
+        Raises:
+            KernelException: If the provided operation is not a subtraction.
+        """
+
+        # check if op is a subtraction
+        if not isinstance(op, arith.Subi):
+            raise KernelException
+
+        # get operands
+        value = op.lhs
+        adjustment = op.rhs
+
+        # additional sign extension possible
+        while isinstance(value, OpResult) and isinstance(value.op, arith.ExtSIOp):
+            value = value.op.operands[0]
+
+        return value, adjustment
+
+    @staticmethod
+    def parse_inputs(linalg_op: linalg.Generic) -> tuple[linalg.YieldOp, dict]:
+        """
+        Parses the inputs of a linalg operation and returns the yield operation
+        of the linalg body and a dictionary of input types. The dictionary of
+        input types is used to match the block arguments (scalars) to the actual
+        input types of the linalg.Generic (shaped memrefs).
+
+        Args:
+            linalg_op (linalg.Generic): The linalg operation.
+
+        Returns:
+            tuple[linalg.YieldOp, dict]: The yield operation and a
+            dictionary of input types.
+
+        Raises:
+            KernelException: If the number of shaped inputs is incorrect for binary ops.
+            KernelException: If the last op of the linalg block is not a yield op.
```

Comment:
I don't think we need to check this here.
To me it seems like this is already checked by the verifier, each `linalg.generic` must be terminated by a `linalg.yield`, since it has the `SingleBlockImplicitTerminator<YieldOp>` trait :)